pkg/statsd/worker.go

```diff
@@ -47,13 +47,21 @@ func (w *worker) executeProcess(cmd *processCommand) {
 }
 
 func (w *worker) RunMetrics(ctx context.Context, statser stats.Statser) {
```

Comment:
This method, like other similar ones that do work concurrently, should block until the goroutine below finishes. `wait.Group` can help here.

Reply:
It does @ash2k , it just forces one to run within the background on a different routine and will cancel once context is done.

Reply:
Unless I have missed something on how contexts work.

Reply:
What I'm saying is this method will return without waiting for `go stats.NewChannelStatsWatcher().Run(ctx)` to terminate. Other places that start goroutines block until the spawned goroutines finish. This is a very important property that allows to write deterministic tests without data races.

Reply:
https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/

Reply:
Would it make sense to put both channels into the `wait.StartWithContext` then wait and the end?
Or just move the singular one there?

Reply:
@MovieStoreGuy something like this:
```
func (w *worker) RunMetrics(ctx context.Context, statser stats.Statser) {
	stgr := stager.New()

	csw := stats.NewChannelStatsWatcher(...chan1..., statser)
	stgr.NextStage().StartWithContext(csw.Run)
	
	csw = stats.NewChannelStatsWatcher(...chan2..., statser)
	stgr.NextStage().StartWithContext(csw.Run)
	
	<-ctx.Done()
	stgr.Shutdown()
}
```

When the context.Done is signaled, stgr.Shutdown() will cancel each stage, in reverse order, and wait for them to complete.  There's a wait.Group behind the scenes which ensures ordering.

Reply:
I ended going for the former since it make the invocation of the watchers the same and neater control flow.

Reply:
Works for me.